File: scripts/serve-static.py
#!/usr/bin/env python3

"""
Static File Server with WebContainer Support (Python Alternative)

Serves static build files from dist/static/ with proper CORS headers
required for WebContainer functionality.

Features:
- WebContainer-compatible CORS headers
- Proper MIME type detection
- SPA fallback routing
- Error handling and logging
- Configurable port

Usage:
    python scripts/serve-static.py [port]
    python scripts/serve-static.py 8080
"""

import os
import sys
import mimetypes
import urllib.parse
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
import signal
import datetime

# Configuration
DEFAULT_PORT = 3000
SCRIPT_DIR = Path(__file__).parent
STATIC_DIR = SCRIPT_DIR.parent / "dist" / "static"
INDEX_FILE = "index.html"

class WebContainerHandler(SimpleHTTPRequestHandler):
    """Custom HTTP handler with WebContainer support"""
    
    def __init__(self, *args, **kwargs):
        # Set the directory to serve from
        super().__init__(*args, directory=str(STATIC_DIR), **kwargs)
    
    def end_headers(self):
        """Add WebContainer-required headers to all responses"""
        # Essential WebContainer headers
        self.send_header('Cross-Origin-Embedder-Policy', 'credentialless')
        self.send_header('Cross-Origin-Opener-Policy', 'same-origin')
        self.send_header('Cross-Origin-Resource-Policy', 'cross-origin')
        
        # Additional CORS headers for broader compatibility
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
        
        # Security headers
        self.send_header('X-Content-Type-Options', 'nosniff')
        self.send_header('X-Frame-Options', 'SAMEORIGIN')
        self.send_header('Referrer-Policy', 'strict-origin-when-cross-origin')
        
        super().end_headers()
    
    def do_GET(self):
        """Handle GET requests with SPA fallback"""
        # Parse the URL
        parsed_path = urllib.parse.urlparse(self.path)
        request_path = parsed_path.path.lstrip('/')
        
        # Default to index.html for root requests
        if not request_path or request_path == '/':
            request_path = INDEX_FILE
        
        # Construct full file path
        file_path = STATIC_DIR / request_path
        
        # Security check: ensure file is within static directory
        try:
            file_path.resolve().relative_to(STATIC_DIR.resolve())
        except ValueError:
            self.log_message("Security warning: Attempted access outside static directory: %s", request_path)
            self.send_error(403, "Forbidden")
            return
        
        # Check if file exists
        if file_path.exists() and file_path.is_file():
            # Serve the file
            self.serve_file(file_path, request_path)
        elif file_path.exists() and file_path.is_dir():
            # Try to serve index.html from directory
            index_path = file_path / INDEX_FILE
            if index_path.exists():
                self.serve_file(index_path, f"{request_path}/{INDEX_FILE}")
            else:
                self.send_error(404, "Directory listing not allowed")
        else:
            # File not found - for SPA, fallback to index.html
            index_path = STATIC_DIR / INDEX_FILE
            if index_path.exists():
                self.log_message("[SPA Fallback] %s -> %s", request_path, INDEX_FILE)
                self.serve_file(index_path, INDEX_FILE)
            else:
                self.send_error(404, "Not Found")
    
    def serve_file(self, file_path, request_path):
        """Serve a file with proper headers"""
        try:
            # Get MIME type
            mime_type, _ = mimetypes.guess_type(str(file_path))
            if mime_type is None:
                mime_type = 'application/octet-stream'
            
            # Read file content
            with open(file_path, 'rb') as f:
                content = f.read()
            
            # Send response
            self.send_response(200)
            self.send_header('Content-Type', mime_type)
            self.send_header('Content-Length', str(len(content)))
            
            # Set cache headers
            if file_path.suffix == '.html':
                self.send_header('Cache-Control', 'no-cache')
            else:
                self.send_header('Cache-Control', 'public, max-age=31536000')
            
            self.end_headers()
            self.wfile.write(content)
            
            # Log successful request
            timestamp = datetime.datetime.now().isoformat()
            self.log_message("[%s] 200 %s (%s)", timestamp, request_path, mime_type)
            
        except Exception as e:
            self.log_error("Error serving file %s: %s", file_path, str(e))
            self.send_error(500, "Internal Server Error")
    
    def do_OPTIONS(self):
        """Handle OPTIONS requests for CORS preflight"""
        self.send_response(200)
        self.end_headers()
    
    def log_message(self, format, *args):
        """Custom log format"""
        sys.stdout.write("%s\n" % (format % args))
        sys.stdout.flush()

def check_static_build():
    """Check if static build exists"""
    if not STATIC_DIR.exists():
        print("❌ Error: Static build directory not found!")
        print(f"   Expected: {STATIC_DIR}")
        print("")
        print("🔧 To create the static build, run:")
        print("   npm run build:static")
        print("   or")
        print("   pnpm build:static")
        print("")
        sys.exit(1)
    
    index_path = STATIC_DIR / INDEX_FILE
    if not index_path.exists():
        print("❌ Error: index.html not found in static build!")
        print(f"   Expected: {index_path}")
        print("")
        print("🔧 Please rebuild the static version:")
        print("   npm run build:static")
        print("")
        sys.exit(1)

def signal_handler(signum, frame):
    """Handle shutdown signals"""
    print("\n🛑 Shutting down server...")
    print("✅ Server stopped successfully")
    sys.exit(0)

def start_server():
    """Start the HTTP server"""
    # Check if static build exists
    check_static_build()
    
    # Get port from command line argument or environment
    port = DEFAULT_PORT
    if len(sys.argv) > 1:
        try:
            port = int(sys.argv[1])
        except ValueError:
            print(f"❌ Error: Invalid port number '{sys.argv[1]}'")
            sys.exit(1)
    elif 'PORT' in os.environ:
        try:
            port = int(os.environ['PORT'])
        except ValueError:
            print(f"❌ Error: Invalid PORT environment variable '{os.environ['PORT']}'")
            sys.exit(1)
    
    # Set up signal handlers for graceful shutdown
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    # Create and start server
    try:
        server = HTTPServer(('localhost', port), WebContainerHandler)
        
        print("🚀 Static File Server with WebContainer Support (Python)")
        print("")
        print(f"📁 Serving: {STATIC_DIR}")
        print(f"🌐 URL: http://localhost:{port}")
        print("")
        print("✅ WebContainer Headers Enabled:")
        print("   • Cross-Origin-Embedder-Policy: credentialless")
        print("   • Cross-Origin-Opener-Policy: same-origin")
        print("   • Cross-Origin-Resource-Policy: cross-origin")
        print("")
        print("🎯 Features:")
        print("   • Static file serving with proper MIME types")
        print("   • SPA fallback routing to index.html")
        print("   • Security headers and CORS support")
        print("   • Request logging and error handling")
        print("")
        print("🛑 To stop the server: Ctrl+C")
        print("")
        
        server.serve_forever()
        
    except OSError as e:
        if e.errno == 48:  # Address already in use
            print(f"❌ Error: Port {port} is already in use!")
            print("   Try a different port:")
            print(f"   python scripts/serve-static.py 8080")
        else:
            print(f"❌ Server error: {e}")
        sys.exit(1)
    except KeyboardInterrupt:
        signal_handler(signal.SIGINT, None)

if __name__ == "__main__":
    start_server()

File: scripts/serve-static.js
#!/usr/bin/env node

/**
 * Static File Server with WebContainer Support
 *
 * Serves static build files from dist/static/ with proper CORS headers
 * required for WebContainer functionality.
 *
 * Features:
 * - WebContainer-compatible CORS headers
 * - Proper MIME type detection
 * - SPA fallback routing
 * - Error handling and logging
 * - Configurable port
 */

import { createServer } from 'http';
import { readFile, stat, readdir } from 'fs/promises';
import { join, extname, dirname } from 'path';
import { fileURLToPath } from 'url';
import { existsSync } from 'fs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Configuration
const DEFAULT_PORT = 3000;
const STATIC_DIR = join(__dirname, '../dist/static');
const INDEX_FILE = 'index.html';

// MIME types mapping
const MIME_TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.htm': 'text/html; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.mjs': 'application/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ttf': 'font/ttf',
  '.eot': 'application/vnd.ms-fontobject',
  '.txt': 'text/plain; charset=utf-8',
  '.xml': 'application/xml; charset=utf-8',
  '.pdf': 'application/pdf',
  '.zip': 'application/zip',
  '.wasm': 'application/wasm',
  '.map': 'application/json; charset=utf-8',
};

/**
 * Get MIME type for file extension
 */
function getMimeType(filePath) {
  const ext = extname(filePath).toLowerCase();
  return MIME_TYPES[ext] || 'application/octet-stream';
}

/**
 * Set WebContainer-required CORS headers
 */
function setWebContainerHeaders(res) {
  // Essential WebContainer headers
  res.setHeader('Cross-Origin-Embedder-Policy', 'credentialless');
  res.setHeader('Cross-Origin-Opener-Policy', 'same-origin');
  res.setHeader('Cross-Origin-Resource-Policy', 'cross-origin');

  // Additional CORS headers for broader compatibility
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  // Security headers
  res.setHeader('X-Content-Type-Options', 'nosniff');
  res.setHeader('X-Frame-Options', 'SAMEORIGIN');
  res.setHeader('Referrer-Policy', 'strict-origin-when-cross-origin');
}

/**
 * Send error response
 */
function sendError(res, statusCode, message) {
  setWebContainerHeaders(res);
  res.writeHead(statusCode, { 'Content-Type': 'text/plain; charset=utf-8' });
  res.end(message);
}

/**
 * Send file response
 */
async function sendFile(res, filePath, requestPath) {
  try {
    const data = await readFile(filePath);
    const mimeType = getMimeType(filePath);

    setWebContainerHeaders(res);
    res.writeHead(200, {
      'Content-Type': mimeType,
      'Content-Length': data.length,
      'Cache-Control': filePath.endsWith('.html') ? 'no-cache' : 'public, max-age=31536000',
    });

    res.end(data);

    // Log successful request
    const timestamp = new Date().toISOString();
    console.log(`[${timestamp}] 200 ${requestPath} (${mimeType})`);
  } catch (error) {
    console.error(`Error serving file ${filePath}:`, error.message);
    sendError(res, 500, 'Internal Server Error');
  }
}

/**
 * Handle incoming requests
 */
async function handleRequest(req, res) {
  const url = new URL(req.url, `http://${req.headers.host}`);
  let requestPath = decodeURIComponent(url.pathname);

  // Remove leading slash and normalize path
  requestPath = requestPath.replace(/^\/+/, '');

  // Default to index.html for root requests
  if (!requestPath || requestPath === '/') {
    requestPath = INDEX_FILE;
  }

  // Construct full file path
  const filePath = join(STATIC_DIR, requestPath);

  // Security check: ensure file is within static directory
  if (!filePath.startsWith(STATIC_DIR)) {
    console.warn(`Security warning: Attempted access outside static directory: ${requestPath}`);
    sendError(res, 403, 'Forbidden');

    return;
  }

  try {
    const stats = await stat(filePath);

    if (stats.isFile()) {
      // Serve the file
      await sendFile(res, filePath, requestPath);
    } else if (stats.isDirectory()) {
      // Try to serve index.html from directory
      const indexPath = join(filePath, INDEX_FILE);

      try {
        await stat(indexPath);
        await sendFile(res, indexPath, `${requestPath}/${INDEX_FILE}`);
      } catch {
        sendError(res, 404, 'Directory listing not allowed');
      }
    } else {
      sendError(res, 404, 'Not Found');
    }
  } catch (error) {
    if (error.code === 'ENOENT') {
      // File not found - for SPA, fallback to index.html
      const indexPath = join(STATIC_DIR, INDEX_FILE);

      try {
        await stat(indexPath);
        console.log(`[SPA Fallback] ${requestPath} -> ${INDEX_FILE}`);
        await sendFile(res, indexPath, INDEX_FILE);
      } catch {
        sendError(res, 404, 'Not Found');
      }
    } else {
      console.error(`Error accessing ${filePath}:`, error.message);
      sendError(res, 500, 'Internal Server Error');
    }
  }
}

/**
 * Check if static build exists
 */
function checkStaticBuild() {
  if (!existsSync(STATIC_DIR)) {
    console.error('❌ Error: Static build directory not found!');
    console.error(`   Expected: ${STATIC_DIR}`);
    console.error('');
    console.error('🔧 To create the static build, run:');
    console.error('   npm run build:static');
    console.error('   or');
    console.error('   pnpm build:static');
    console.error('');
    process.exit(1);
  }

  const indexPath = join(STATIC_DIR, INDEX_FILE);

  if (!existsSync(indexPath)) {
    console.error('❌ Error: index.html not found in static build!');
    console.error(`   Expected: ${indexPath}`);
    console.error('');
    console.error('🔧 Please rebuild the static version:');
    console.error('   npm run build:static');
    console.error('');
    process.exit(1);
  }
}

/**
 * Start the server
 */
function startServer() {
  // Check if static build exists
  checkStaticBuild();

  // Get port from environment or use default
  const port = process.env.PORT || DEFAULT_PORT;

  // Create HTTP server
  const server = createServer(handleRequest);

  // Handle server errors
  server.on('error', (error) => {
    if (error.code === 'EADDRINUSE') {
      console.error(`❌ Error: Port ${port} is already in use!`);
      console.error('   Try a different port:');
      console.error(`   PORT=8080 npm run serve:static`);
    } else {
      console.error('❌ Server error:', error.message);
    }

    process.exit(1);
  });

  // Start listening
  server.listen(port, () => {
    console.log('🚀 Static File Server with WebContainer Support');
    console.log('');
    console.log(`📁 Serving: ${STATIC_DIR}`);
    console.log(`🌐 URL: http://localhost:${port}`);
    console.log('');
    console.log('✅ WebContainer Headers Enabled:');
    console.log('   • Cross-Origin-Embedder-Policy: credentialless');
    console.log('   • Cross-Origin-Opener-Policy: same-origin');
    console.log('   • Cross-Origin-Resource-Policy: cross-origin');
    console.log('');
    console.log('🎯 Features:');
    console.log('   • Static file serving with proper MIME types');
    console.log('   • SPA fallback routing to index.html');
    console.log('   • Security headers and CORS support');
    console.log('   • Request logging and error handling');
    console.log('');
    console.log('🛑 To stop the server: Ctrl+C');
    console.log('');
  });

  // Graceful shutdown
  process.on('SIGINT', () => {
    console.log('\n🛑 Shutting down server...');
    server.close(() => {
      console.log('✅ Server stopped successfully');
      process.exit(0);
    });
  });

  process.on('SIGTERM', () => {
    console.log('\n🛑 Received SIGTERM, shutting down...');
    server.close(() => {
      console.log('✅ Server stopped successfully');
      process.exit(0);
    });
  });
}

// Start the server
startServer();
